File: src/main/java/com/mercateo/jsonschema/property/AnnotationMapBuilder.java
package com.mercateo.jsonschema.property;

import java.lang.annotation.Annotation;

import com.google.common.collect.HashMultimap;
import com.google.common.collect.Multimap;

class AnnotationMapBuilder {
    Multimap<Class<? extends Annotation>, Annotation> createMap(
            Annotation[] annotations) {
        final Multimap<Class<? extends Annotation>, Annotation> annotationMap = HashMultimap
                .create();
        for (Annotation annotation : annotations) {
            annotationMap.put(annotation.annotationType(), annotation);
        }
        return annotationMap;
    }

    Multimap<Class<? extends Annotation>, Annotation> merge(
            Multimap<Class<? extends Annotation>, Annotation> annotations,
            Multimap<Class<? extends Annotation>, Annotation> otherAnnotations) {

        final HashMultimap<Class<? extends Annotation>, Annotation> mergedAnnotations = HashMultimap.create(
                annotations);
        mergedAnnotations.putAll(otherAnnotations);
        return mergedAnnotations;
    }
}

File: src/main/java/com/mercateo/jsonschema/property/AnnotationMapBuilder.kt
package com.mercateo.jsonschema.property

import java.util.*

class AnnotationMapBuilder {
    fun createMap(vararg annotations: Annotation): Map<Class<out Annotation>, Set<Annotation>> {
        return annotations.groupBy({ it -> it.annotationClass.java }, { it }).mapValues { it.value.toSet() }
    }

    fun merge(annotations: Map<Class<out Annotation>, Set<Annotation>>,
              otherAnnotations: Map<Class<out Annotation>, Set<Annotation>>): Map<Class<out Annotation>, Set<Annotation>> {

        return annotations.mergeReduce(otherAnnotations, { obj, elements -> obj + elements })
    }

    private fun <K, V> Map<K, V>.mergeReduce(other: Map<K, V>, reduce: (V, V) -> V = { a, b -> b }): Map<K, V> {
        val result = HashMap<K, V>(this.size + other.size)
        result.putAll(this)
        other.forEach { e ->
            val existing = result[e.key]

            if (existing == null) {
                result[e.key] = e.value
            } else {
                result[e.key] = reduce(e.value, existing)
            }
        }

        return result
    }
}
